iavl_nodedb.go

```diff
@@ -73,68 +107,418 @@ func (ndb *nodeDB) SaveNode(node *IAVLNode) {
+	}
+}
+
+// Saves orphaned nodes to disk under a special prefix.
+func (ndb *nodeDB) SaveOrphans(version uint64, orphans map[string]uint64) {
+	ndb.mtx.Lock()
+	defer ndb.mtx.Unlock()
+
+	toVersion := ndb.getPreviousVersion(version)
+
+	for hash, fromVersion := range orphans {
+		ndb.saveOrphan([]byte(hash), fromVersion, toVersion)
+	}
+}
+
+// Saves a single orphan to disk.
+func (ndb *nodeDB) saveOrphan(hash []byte, fromVersion, toVersion uint64) {
+	if fromVersion > toVersion {
+		cmn.PanicSanity("Orphan expires before it comes alive")
+	}
+	key := fmt.Sprintf(orphansKeyFmt, toVersion, fromVersion, hash)
+	ndb.batch.Set([]byte(key), hash)
+
+	// Set reverse-lookup index.
+	indexKey := fmt.Sprintf(orphansIndexKeyFmt, hash)
+	ndb.batch.Set([]byte(indexKey), []byte(key))
+}
+
+// deleteOrphans deletes orphaned nodes from disk, and the associated orphan
+// entries.
+func (ndb *nodeDB) deleteOrphans(version uint64) {
+	// Will be zero if there is no previous version.
+	predecessor := ndb.getPreviousVersion(version)
+
+	// Traverse orphans with a lifetime ending at the version specified.
+	ndb.traverseOrphansVersion(version, func(key, hash []byte) {
+		var fromVersion, toVersion uint64
+
+		// See comment on `orphansKeyFmt`. Note that here, `version` and
+		// `toVersion` are always equal.
+		fmt.Sscanf(string(key), orphansKeyFmt, &toVersion, &fromVersion)
+
+		// Delete orphan key and reverse-lookup key.
+		ndb.batch.Delete(key)
+		ndb.batch.Delete([]byte(fmt.Sprintf(orphansIndexKeyFmt, hash)))
+
+		// If there is no predecessor, or the predecessor is earlier than the
+		// beginning of the lifetime (ie: negative lifetime), or the lifetime
+		// spans a single version and that version is the one being deleted, we
+		// can delete the orphan.  Otherwise, we shorten its lifetime, by
+		// moving its endpoint to the previous version.
+		if predecessor < fromVersion || fromVersion == toVersion {
+			ndb.batch.Delete(ndb.nodeKey(hash))
+			ndb.uncacheNode(hash)
+		} else {
+			ndb.saveOrphan(hash, fromVersion, predecessor)
+		}
+	})
+}
+
+func (ndb *nodeDB) nodeKey(hash []byte) []byte {
+	return []byte(fmt.Sprintf(nodesKeyFmt, hash))
+}
+
+func (ndb *nodeDB) getLatestVersion() uint64 {
+	if ndb.latestVersion == 0 {
+		ndb.getVersions()
+	}
+	return ndb.latestVersion
+}
+
+func (ndb *nodeDB) getVersions() map[uint64][]byte {
+	if len(ndb.versionCache) == 0 {
+		ndb.traversePrefix([]byte(rootsPrefix), func(k, hash []byte) {
+			var version uint64
+			fmt.Sscanf(string(k), rootsPrefixFmt, &version)
+			ndb.cacheVersion(version, hash)
+		})
+	}
+	return ndb.versionCache
+}
+
+func (ndb *nodeDB) cacheVersion(version uint64, hash []byte) {
+	ndb.versionCache[version] = hash
+
+	if version > ndb.getLatestVersion() {
+		ndb.latestVersion = version
+	}
+}
+
+func (ndb *nodeDB) getPreviousVersion(version uint64) uint64 {
+	var result uint64 = 0
+	for v, _ := range ndb.getVersions() {
+		if v < version && (result == 0 || v > result) {
```

Comment:
is `result == 0` necessary ?

Reply:
Also maybe there's a more efficient way to keep the previous version around so we dont have to iterate a map?